cmake/alpakaCommon.cmake

```diff
@@ -0,0 +1,881 @@
+                        if(CMAKE_CXX_COMPILER_VERSION GREATER_EQUAL 9.0)
+                            message(FATAL_ERROR "NVCC 10.1 does not support clang 9+. Please use clang 4, 5, 6, 7 or 8!")
+                        endif()
+                    elseif(CUDA_VERSION VERSION_EQUAL 10.2)
+                        if(CMAKE_CXX_COMPILER_VERSION GREATER_EQUAL 9.0)
+                            message(FATAL_ERROR "NVCC 10.2 does not support clang 9+. Please use clang 4, 5, 6, 7 or 8!")
+                        endif()
+                    endif()
+                endif()
+
+                if(ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE)
+                    message(FATAL_ERROR "NVCC does not support boost.fiber!")
+                endif()
+
+                # Clean up the flags. Else, multiple find calls would result in duplicate flags. Furthermore, other modules may have set different settings.
+                set(CUDA_NVCC_FLAGS)
+
+                if(${ALPAKA_DEBUG} GREATER 1)
+                    set(CUDA_VERBOSE_BUILD ON)
+                endif()
+
+                set(CUDA_PROPAGATE_HOST_FLAGS ON)
+
+                if(ALPAKA_CUDA_NVCC_SEPARABLE_COMPILATION)
+                    set(CUDA_SEPARABLE_COMPILATION ON)
+                endif()
+
+                # nvcc sets no linux/__linux macros on OpenPOWER linux
+                # nvidia bug id: 2448610
+                if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+                    if(CMAKE_SYSTEM_PROCESSOR STREQUAL "ppc64le")
+                        list(APPEND CUDA_NVCC_FLAGS "-Dlinux")
+                    endif()
+                endif()
+
+                # NOTE: Since CUDA 10.2 this option is also alternatively called '--extended-lambda'
+                if(ALPAKA_CUDA_NVCC_EXPT_EXTENDED_LAMBDA)
+                    list(APPEND CUDA_NVCC_FLAGS "--expt-extended-lambda")
+                endif()
+                # This is mandatory because with c++14 many standard library functions we rely on are constexpr (std::min, std::multiplies, ...)
+                list(APPEND CUDA_NVCC_FLAGS "--expt-relaxed-constexpr")
+
+                foreach(_CUDA_ARCH_ELEM ${ALPAKA_CUDA_ARCH})
+                    # set flags to create device code for the given architecture
+                    list(APPEND CUDA_NVCC_FLAGS
+                        --generate-code arch=compute_${_CUDA_ARCH_ELEM},code=sm_${_CUDA_ARCH_ELEM}
+                        --generate-code arch=compute_${_CUDA_ARCH_ELEM},code=compute_${_CUDA_ARCH_ELEM}
+                    )
+                endforeach()
+
+                if(NOT MSVC)
+                    list(APPEND CUDA_NVCC_FLAGS "-std=c++${ALPAKA_CXX_STANDARD}")
+                endif()
+
+                set(CUDA_HOST_COMPILER "${CMAKE_CXX_COMPILER}")
+
+                list(APPEND CUDA_NVCC_FLAGS $<$<OR:$<CONFIG:Debug>, $<CONFIG:RelWithDebInfo>>:"-g">)
+                list(APPEND CUDA_NVCC_FLAGS $<$<AND:$<OR:$<CONFIG:DEBUG>, $<CONFIG:RelWithDebInfo>>, $<NOT$<CXX_COMPILER_ID:MSVC>>>:"-G">)
```

Comment:
This results in:
```
CMake Error at /home/travis/CMake/share/cmake-3.16/Modules/FindCUDA.cmake:1603 (file):
  Error evaluating generator expression:
    $<CXX_COMPILER_ID:MSVC>
  $<CXX_COMPILER_ID> may only be used with binary targets.  It may not be
  used with add_custom_command or add_custom_target.
Call Stack (most recent call first):
  /home/travis/CMake/share/cmake-3.16/Modules/FindCUDA.cmake:1868 (CUDA_WRAP_SRCS)
  cmake/addExecutable.cmake:37 (CUDA_ADD_EXECUTABLE)
  example/bufferCopy/CMakeLists.txt:45 (alpaka_add_executable)
```

Reply:
Ugh, so we can't use generator expressions here. Back to if-else then.